stdlib/public/Concurrency/Actor.cpp

```diff
@@ -1357,21 +1359,66 @@ void swift::swift_task_enqueue(Job *job, ExecutorRef executor) {
   if (executor.isDefaultActor())
     return asImpl(executor.getDefaultActor())->enqueue(job);
 
+  // Just assume it's actually a default actor that we haven't tagged
+  // properly.
   // FIXME: call the general method.
-  job->run(executor);
+  return asImpl(reinterpret_cast<DefaultActor*>(executor.getRawValue()))
+    ->enqueue(job);
 }
 
 SWIFT_CC(swift)
 void (*swift::swift_task_enqueueGlobal_hook)(Job *job) = nullptr;
 
+static void __swift_run_job(void *_job) {
+  Job *job = (Job*) _job;
+  job->run(ExecutorRef::generic());
+}
+
 void swift::swift_task_enqueueGlobal(Job *job) {
   assert(job && "no job provided");
 
   // If the hook is defined, use it.
   if (swift_task_enqueueGlobal_hook)
     return swift_task_enqueueGlobal_hook(job);
 
-  // FIXME: implement this properly
-  job->run(ExecutorRef::generic());
+  // We really want four things from the global execution service:
+  //  - Enqueuing work should have minimal runtime and memory overhead.
+  //  - Adding work should never result in an "explosion" where many
+  //    more threads are created than the available cores.
+  //  - Jobs should run on threads with an appropriate priority.
+  //  - Thread priorities should temporarily elevatable to avoid
+  //    priority inversions.
+  //
+  // Of these, the first two are the most important.  Many programs
+  // do not rely on high-usage priority scheduling, and many priority
+  // inversions can be avoided at a higher level (albeit with some
+  // performance cost, e.g. by creating higher-priority tasks to run
+  // critical sections that contend with high-priority work).  In
+  // contrast, if the async feature adds too much overhead, or if
+  // heavy use of it leads to thread explosions and memory exhaustion,
+  // programmers will have no choice but to stop using it.  So if
+  // goals are in conflict, it's best to focus on core properties over
+  // priority-inversion avoidance.
+
+  // We currently use Dispatch for our thread pool on all platforms.
+  // Dispatch currently backs its serial queues with a global
+  // concurrent queue that is prone to thread explosions when a flood
+  // of jobs are added to it.  That problem does not apply equally
+  // to the global concurrent queues returned by dispatch_get_global_queue,
+  // which are not strictly CPU-limited but are at least much more
+  // cautious about adding new threads.  We cannot safely elevate
+  // the priorities of work added to this queue using Dispatch's public
+  // API, but as discussed above, that is less important than avoiding
+  // performance problems.
+  dispatch_function_t dispatchFunction = &__swift_run_job;
+  void *dispatchContext = job;
+
+  JobPriority priority = job->getPriority();
+
+  // TODO: cache this to avoid the extra call
+  auto queue = dispatch_get_global_queue((qos_class_t) priority,
```

Comment:
As I noted elsewhere, I think this needs to be `dispatch_qos_class_t`.

Reply:
Yeah, I saw that and have the fix already; I'll update the PR as soon as this round of build comes back.